src/placement/mapless_jump_map.c

```diff
@@ -0,0 +1,998 @@
+/**
+ *
+ *
+ * (C) Copyright 2016-2019 Intel Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. B609815.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+/**
+ * This file is part of DSR
+ *
+ * src/placement/mapless_jump_map.c
+ */
+#define D_LOGFAC        DD_FAC(placement)
+
+#include "pl_map.h"
+#include <inttypes.h>
+#include <daos/pool_map.h>
+
+/**
+ * This struct is used to to hold information while
+ * finding rebuild targets for shards located on unavailable
+ * targets.
+ */
+struct remap_node {
+	uint32_t rank;
+	uint32_t shard_idx;
+	uint32_t fseq;
+};
+
+
+/**
+ * Helper function used for sorting rebuild list by fail sequence
+ * This function compares the fail sequence of the remap_nodes
+ */
+static int
+map_rebuild_cmp(void *array, int a, int b)
+{
+	struct remap_node *sort = (struct remap_node *)array;
+
+	if (sort[a].fseq > sort[b].fseq)
+		return 1;
+	if (sort[a].fseq  < sort[b].fseq)
+		return -1;
+	return 0;
+}
+
+/**
+ * Helper function used for sorting rebuild list by fail sequence
+ * This function swaps the two remap_nodes.
+ */
+static void
+map_rebuild_swap(void *array, int a, int b)
+{
+	struct remap_node *sort;
+	struct remap_node  tmp;
+
+	sort = ((struct remap_node *)array);
+
+	tmp = sort[a];
+	sort[a] = sort[b];
+	sort[b] = tmp;
+}
+
+/** Function pointers for sorting the rebuild list. */
+static daos_sort_ops_t map_rebuild_sops = {
+	.so_cmp		= map_rebuild_cmp,
+	.so_swap	= map_rebuild_swap,
+};
+
+/**
+ * This function sets a specific bit in the bitmap
+ * It expects the bitmap to be zero indexed from left
+ * to right.
+ */
+static inline void
+set_bit(uint8_t *bitmap, uint64_t bit)
+{
+	uint64_t offset = bit / 8;
+	uint8_t position = bit % 8;
+	bitmap[offset] |= (0x80 >> position);
+}
+
+/** Returns the bit at a specific position in the bitmap */
+static inline uint8_t
+get_bit(uint8_t *bitmap, uint64_t bit)
+{
+	uint64_t offset = bit / 8;
+        uint8_t position = bit % 8;
+        return ((bitmap[offset] & (0x80 >> position)) != 0);
```

Comment:
(style)  code indent should use tabs where possible
(style)  Missing a blank line after declarations